packages/xai/src/responses/xai-responses-api.ts

```diff
@@ -0,0 +1,211 @@
+  role: 'system' | 'developer';
+  content: string;
+};
+
+export type XaiResponsesUserMessage = {
+  role: 'user';
+  content: string;
+};
+
+export type XaiResponsesAssistantMessage = {
+  role: 'assistant';
+  content: string;
+};
+
+export type XaiResponsesTool =
+  | {
+      type: 'web_search';
+      allowed_domains?: string[];
+      excluded_domains?: string[];
+      enable_image_understanding?: boolean;
+    }
+  | {
+      type: 'x_search';
+      allowed_x_handles?: string[];
+      excluded_x_handles?: string[];
+      from_date?: string;
+      to_date?: string;
+      enable_image_understanding?: boolean;
+      enable_video_understanding?: boolean;
+    }
+  | { type: 'code_interpreter' }
+  | { type: 'file_search' }
+  | { type: 'mcp' }
+  | {
+      type: 'function';
+      function: {
+        name: string;
+        description?: string;
+        parameters: unknown;
+      };
+    };
+
+const annotationSchema = z.union([
+  z.object({
+    type: z.literal('url_citation'),
+    url: z.string(),
+    title: z.string().optional(),
+  }),
+  z.object({
+    type: z.string(),
+  }),
+]);
+
+const messageContentPartSchema = z.object({
+  type: z.string(),
+  text: z.string().optional(),
+  logprobs: z.array(z.any()).optional(),
+  annotations: z.array(annotationSchema).optional(),
+});
+
+const toolCallSchema = z.object({
+  name: z.string(),
+  arguments: z.string(),
+  call_id: z.string(),
+  id: z.string(),
+  status: z.string(),
+});
+
+const outputItemSchema = z.discriminatedUnion('type', [
+  z.object({
+    type: z.literal('web_search_call'),
+    ...toolCallSchema.shape,
+  }),
+  z.object({
+    type: z.literal('x_search_call'),
+    ...toolCallSchema.shape,
+  }),
+  z.object({
+    type: z.literal('code_interpreter_call'),
+    ...toolCallSchema.shape,
+  }),
+  z.object({
+    type: z.literal('code_execution_call'),
+    ...toolCallSchema.shape,
+  }),
+  z.object({
+    type: z.literal('view_image_call'),
+    ...toolCallSchema.shape,
+  }),
+  z.object({
+    type: z.literal('view_x_video_call'),
+    ...toolCallSchema.shape,
+  }),
+  z.object({
+    type: z.literal('message'),
+    role: z.string(),
+    content: z.array(messageContentPartSchema),
+    id: z.string(),
+    status: z.string(),
+  }),
+  z.object({
+    type: z.literal('function_call'),
+    name: z.string(),
+    arguments: z.string(),
+    call_id: z.string(),
+    id: z.string(),
+  }),
+]);
+
+export const xaiResponsesUsageSchema = z.object({
+  input_tokens: z.number(),
+  output_tokens: z.number(),
+  total_tokens: z.number().optional(),
+  input_tokens_details: z
+    .object({
+      cached_tokens: z.number().optional(),
+    })
+    .optional(),
+  output_tokens_details: z
+    .object({
+      reasoning_tokens: z.number().optional(),
+    })
+    .optional(),
+  num_sources_used: z.number().optional(),
+  num_server_side_tools_used: z.number().optional(),
+});
+
+export const xaiResponsesResponseSchema = z.object({
```

Comment:
we should use lazySchema, see https://github.com/vercel/ai/issues/9339. You can do it in a follow up PR though and do all of xAI then.